File: 0075-sort-colors/0075-sort-colors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int zero = 0, one = 0, two = 0;
        int n = nums.size();
        
        // Count occurrences of 0, 1, and 2
        for (int i = 0; i < n; i++) {
            if (nums[i] == 0) zero++;
            else if (nums[i] == 1) one++;
            else if (nums[i] == 2) two++;
        }
        
        // Assign 0s, 1s, and 2s back to the array
        int j = 0;
        for (int i = 0; i < zero; i++) nums[j++] = 0;
        for (int i = 0; i < one; i++) nums[j++] = 1;
        for (int i = 0; i < two; i++) nums[j++] = 2;
    }
};
/*public:
    void sortColors(vector<int>& nums) {
       int low = 0;
       int mid =0;
       int high = nums.size()-1;
       while(mid <= high){
         if(nums[mid]==0){
            swap(nums[mid],nums[low]);
            low++;
            mid++;
         }
         else if(nums[mid]==2){
            swap(nums[mid],nums[high]);
            high--;
         }
         else{
            mid++;
         }
       }
     
    }*/

File: 0075-sort-colors/0075-sort-colors.java
class Solution {
    
    private void swap(int[] nums, int start, int end){
        int temp = nums[start];

        nums[start] = nums[end];
        nums[end] = temp;
    }
    public void sortColors(int[] nums) {
        int n = nums.length;
        int low=0,mid=0,high = nums.length-1;

        //dutch flag algorithm
        while(mid <= high){

            if(nums[mid] ==0){
                swap(nums,mid,low);
                mid++;
                low++; //to maintain 0 to l-1 as 0;
            }
            else if(nums[mid] ==1){
               
                mid++; //to maintain 1 between low to mid-1;
            }
            else{
                swap(nums,high,mid);
                high--; //to maintain the mid to high as unsorted
            }
        }
    }
}